### PR TITLE
Fix T5 kv cache

### PR DESCRIPTION
**Pull request by jbochi**
This fixes the issue I mentioned in https://github.com/huggingface/candle/pull/892 and https://github.com/huggingface/candle/pull/864#issuecomment-1724176512

The position biases were not taking into account the cache.

I now get the same output with the two approaches, but somehow, it's slightly slower when the cache is on.  (I have a M2 Mac. I didn't try with a GPU, larger batches, or much larger outputs.)

```bash
$ cargo run --release --example t5 -- --model-id "t5-base" --prompt "summarize: state authorities dispatched emergency crews tuesday to survey the damage after an onslaught of severe weather in mississipi." --temperature 0 --decode
    Finished release [optimized] target(s) in 0.18s
     Running `target/release/examples/t5 --model-id t5-base --prompt 'summarize: state authorities dispatched emergency crews tuesday to survey the damage after an onslaught of severe weather in mississipi.' --temperature 0 --decode`
Running on CPU, to run on GPU, build this example with `--features cuda`
 mississippi authorities dispatch emergency crews to survey damage . severe weather in mississippi has caused extensive damage .
38 tokens generated (15.90 token/s)
```

```bash
$ cargo run --release --example t5 -- --model-id "t5-base" --prompt "summarize: state authorities dispatched emergency crews tuesday to survey the damage after an onslaught of severe weather in mississipi." --temperature 0 --decode --disable-cache
    Finished release [optimized] target(s) in 0.17s
     Running `target/release/examples/t5 --model-id t5-base --prompt 'summarize: state authorities dispatched emergency crews tuesday to survey the damage after an onslaught of severe weather in mississipi.' --temperature 0 --decode --disable-cache`
Running on CPU, to run on GPU, build this example with `--features cuda`
 mississippi authorities dispatch emergency crews to survey damage . severe weather in mississippi has caused extensive damage .
38 tokens generated (17.94 token/s)
```

### CODE REVIEW

File: candle-transformers/src/models/t5.rs [hunk 1]
```diff
@@ -348,9 +348,14 @@ impl T5Attention {
                 None => (scores, None),
                 Some(relative_attention_bias) => {
                     // This only handles the bidirectional case.
+                    let kv_len = k.dim(2)?;
```
Comment (LaurentMazare) [inline candle-transformers/src/models/t5.rs:351]: This doesn't seem right to me for the case where `use_cache=true` and the query size is larger than 1 (typically the case when processing a prompt). I would think it's more something like `kv_len - q_len` to `kv_len`?

**Review by LaurentMazare — COMMENTED** [review]
Nice, thanks for taking a look into this.

File: candle-transformers/src/models/t5.rs (hunk 1 above)
Comment (jbochi) [reply candle-transformers/src/models/t5.rs:351]: Sorry, I'm not sure if I follow.

I added a new `decoder_prompt` to get to this scenario where `use_cache=true` and query size is larger than 1 in both the encoder and decoder and it works fine:

```bash
$ cargo run --release --example t5 -- --model-id "t5-base" --prompt "summarize: state authorities dispatched emergency crews tuesday to survey the damage after an onslaught of severe weather in mississipi." --decoder-prompt "the weather has" --temperature 0 --decode
    Finished release [optimized] target(s) in 0.17s
     Running `target/release/examples/t5 --model-id t5-base --prompt 'summarize: state authorities dispatched emergency crews tuesday to survey the damage after an onslaught of severe weather in mississipi.' --decoder-prompt 'the weather has' --temperature 0 --decode`
Running on CPU, to run on GPU, build this example with `--features cuda`
the weather has been severe in mississipi . state authorities dispatched emergency crews to survey the damage .
29 tokens generated (15.39 token/s)
```

Comment (LaurentMazare) [reply candle-transformers/src/models/t5.rs:351]: Isn't the `q_start=kv_len-1`, and `q_end=kv_len` assuming a query length of 1?

Comment (jbochi) [reply candle-transformers/src/models/t5.rs:351]: (ps: this code is never executed for cross-attention, so the variable names may be a bit misleading)

Comment (jbochi) [reply candle-transformers/src/models/t5.rs:351]: Oh, I see what you mean, sorry. I was looking at the line 351, not 352.

Not sure how it was working before 🤔 . Fixed it now.

**Review by LaurentMazare — APPROVED** [review]

**Comment by LaurentMazare** [comment]
Thanks again!

**Pull request merged**